Cookbook/Proposals/MocksAndStubsNextToImplementation.md

```diff
@@ -0,0 +1,53 @@
+# Put Mocks, Stubs Next to Implementation
+
+* Author(s): Sergey Shulga
+* Review Manager: TBD
+
+## Introduction
+
+Put implementation of mocks next to the implementation of the protocol
+
+## Motivation
+
+
+- Improves Discoverability of mocks
+- Prevents mocks duplication (Because we usually define mocks in the `FrameworkTests` target which means they can only be used there)
+- Easy to change mock implementation once public API changes
+
+
+## Proposed solution
+
+
+```swift
+public protocol BiometricAuthControllerProtocol {
+    var isRegistered: Bool { get }
+    var supportedBiometry: BiometryType { get }
+    var biometricDataHasChanged: Bool { get }
+    
+    func register() -> SignalProducer<Never, BiometricAuthError>
+    func unregister() -> SignalProducer<Never, BiometricAuthError>
+    func signedNonce() -> SignalProducer<Data, BiometricAuthError>
+}
+
+public final class BiometricAuthController: BiometricAuthControllerProtocol {
+// Implementation
+}
+
+#if DEBUG
```

Comment:
This should be a separate pre-compiler directive if we accept the proposal. There exist scenarios when we want to run unit tests with optimisations enable. My suggestion would be `#if DEBUG` -> `#if UNITTEST`.

Reply:
I'm not sold on the idea to have different compilation condition. It will force us to create a new build configuration for each and every target unless there is a better way.
Regarding having mock types in the same file I think it will be better to keep them separate from production code and close to test cases, so in the `Tests` subfolder

Reply:
> Tests subfolder
That defeats the purpose of discoverability. You should be able to go the the Interface and straight a way see the implementation and mocks

Reply:
I agree with that, but still not sure about having mocks code right there is. How do we ensure that we don't forget to wrap this code with `#if ... #end`? Or we don't care a lot if we accidentally ship it?

Reply:
If keeping mocks close to the source files is required, they could also be in a separate files but right next to the source file with a ```XXX+Mock.swift```. I suggest suffixes so that alphabetically they are as close to the source file as possible. But the mock file itself IMHO should only be in the test target else we could end up in a lot of trouble by accidentally shipping mock files in production code.

Reply:
I like `Mock`s the `Tests` subfolder and it is an improvement considering it is in the same target. 
In the same file is very risky. Another file prefixed with `Mock` is more explicit.

Reply:
So @ilyapuchka your main concern is Mocks accidentally getting shipped? And by having them in a separate file, they can be added to a separate target? The likelihood of forgetting to add `#if DEBUG` is pretty small IMO. And you could just as easily add the mock file to the main target instead of the test target by mistake and have the same problem.

Reply:
1. The problem is not just that but also the fact that we can break release build having debug build perfectly fine, this happened not so long time ago and it's hard to spot until we build for release which we don't do constantly.
With suggested approach of using excluded files build setting to filter out fixture files (which IIRC we agreed on on the proposal meeting) we only need to maintain file name convention (and we can use regexp I think) to ensure that fixtures are being excluded. I saw we already follow this convention of putting fixtures in `+fixture.swift` files, but probably not everywhere.

2. Having tests related files and actual production code all mixed together in single folder and in single group in Xcode also is not great IMO as it's harder to navigate project structure this way. And we agreed to put all tests into Tests subfolder, so fixtures can go there as well. One can argue that fixtures in this Tests folder will belong to framework target and tests will actually belong to tests target, which may be confusing and people can consider that an error. Then we can have a different sub folder just for Fixtures. 

3. It also makes it potentially easier to analyse what changed in the automation if we keep source code separate from tests. And if ever we will decide to move to SPM to manage internal dependencies it will be easier to migrate (but it will invalidate the whole approach of having Tests sub folder in the sources folder).